### PR TITLE
Add and remove highlights, plus some fixes

### LINKED ISSUE #58: Fragments aren't drawn in the correct tab
### Steps to reproduce

- Open a known page in a first tab
- While LaSuli loads fragments for this page, go to a second tab
- When LaSuli finished collecting the fragments, it displays them in the current (second, wrong) tab

### How to fix

(Warning: unreadable technical stuff).

Injection of the highlighting script is done by the `Display` component (`src/sidebar/jsx/Display.jsx`), in order to fulfill the requirement "show highlights whenever the sidebar is open". The private `_loadScript()` method asks the browser for the current active tab. When the current active tab is not the good one, this bug happens.

Loading the `Display` component is therefore done by the `Sidebar` component (`src/sidebar/sidebar.jsx`), in the `getContent()` method, called for every `this.setState({…});` occurrence.

The one we're interested in is located near the end of the `_updateContent()` private method, as a callback to the `getResource` browser call (to the `model` script).

There is a tab change listener in the `componentDidMount()` React method. I will dig into it this weekend, but a fix would be to find a non-ugly way to store the tab id in `_updateContent()` to pass it to `Display` (why not through the component's state).

### PR DESCRIPTION
**Pull request by franck-eyraud**
Implements 
#47 #52 

Fixes
#58 #60

### CODE REVIEW

File: src/backgroundScripts/background.js [hunk 1]
```diff
@@ -74,3 +81,26 @@ tabs.onActivated.addListener(async (activeInfo) => {
...
+		return model.createHighlight(msg.uri,msg.viewpoint,msg.topic,msg.coordinates);
+	}
+	if (msg.aim === 'removeHighlight') {
+		return model.removeHighlight(msg.uri,msg.viewpoint,msg.topic,msg.fragId);
+	}
+	if (msg.aim === 'setHLNumber') {
+		if (msg.count && msg.tabId) {
+			return setHLNumber(msg.count,msg.tabId);
+		}
+	}
```
Comment (benel) [inline src/backgroundScripts/background.js:102]: Why not a switch/case structure?

**Comment by benel** [comment]
Looks great. I will test it.

**Comment by benel** [comment]
I've got a problem when I'm authenticated on the highlighted page:
false auth

I'm not sure about the problem (maybe the credentials are forwarded from the page to the scripts?).

You can test it on `/text/IF14/b61bc2c460f4a9f60953756797a81d88` with your personal Hypertopic account.

**Comment by benel** [comment]
I understand. The bug is because one of the request (the one to Cassandre) got a 401.
Maybe the process should continue when one of the server is successful.

File: src/backgroundScripts/background.js (hunk 1 above)
Comment (franck-eyraud) [reply src/backgroundScripts/background.js:102]: You are right, it should be. This is a mere move of a part of code in another file, which extended an [existing](https://github.com/Hypertopic/LaSuli/pull/61/commits/7c8dc841b371f4bde66a8d14933e55e998b69adf#diff-2b6e46436cc8167726516dc3bacdf6acL247) `if` series, and I didn't transform it in a switch/case block.
I'll commit it, not before tomorrow.

**Comment by franck-eyraud** [comment]
> The bug is because one of the request (the one to Cassandre) got a 401. Maybe the process should continue when one of the server is successful.

I'll have a look. Do you think that this problem was introduced by these changes, or already existed ?

**Comment by benel** [comment]
> Do you think that this problem was introduced by these changes, or already existed ?

I'm not sure... Maybe I should handle the errors differently in Hypertopic node library. I use `Promise.all`:

https://github.com/Hypertopic/lib-node/blob/master/src/index.js#L16

**Comment by benel** [comment]
Maybe we could replace the faulty line in Hypertopic node lib getView:

```javascript
 return Promise.all(uris.map(_get))
```

with: 

```javascript
 return Promise.all(uris.map(_safeGet))
```

where `_safeGet` return `{rows:[]}` on HTTP errors.

**Comment by benel** [comment]
Hmmm. In fact `_get` is only used in `getView` and it doesn't throw any exception.
More tests have to be done.

**Comment by benel** [comment]
In fact, the problem is not the 401 status itself but the body (`Unauthorized`) that is not valid JSON.
I'll try to fix this.

**Comment by benel** [comment]
OK. Seems to work. I will publish an updated version of the library.

**Comment by benel** [comment]
The following code: 

```javascript
const hypertopic = require('hypertopic');

var db = hypertopic([
  "http://argos2.hypertopic.org", 
  "http://cassandre.hypertopic.org"
]);

var _log = function _log(x) {
  return console.log(JSON.stringify(x, null, 2));
};

db.getView('/item/?resource=http://cassandre.hypertopic.org/text/IF14/b61bc2c460f4a9f60953756797a81d88')
  .then(_log);
```

displays (with the new 3.1.5 version of Hypertopic/lib-node): 
```json
{
  "http://cassandre.hypertopic.org/text/IF14/b61bc2c460f4a9f60953756797a81d88": {
    "item": [
      {
        "corpus": "b61bc2c460f4a9f60953756797c36361",
        "id": "b61bc2c460f4a9f60953756797c36820"
      }
    ]
  }
}
```

**Comment by benel** [comment]
Here is the updated part of the library: https://github.com/Hypertopic/lib-node/commit/d9ca553341d9ef3a9610dbfcf39962173a123923.

**Comment by franck-eyraud** [comment]
Thank you for this analysis. This probably handles the issue. 
However, what is strange is that I tested this page without upgrading the Hypertopic library (with both argos and argos.test servers in Lasuli configuration) but I never got this error. Were you getting it when loading the highlights ?

However, if one is not identified in the argos server, then any modification is currently silently ignored (except the 401 unauthorized displayed in the web browser console), and there it would be probably nice to display some message. Do you have some idea where ?

File: src/backgroundScripts/background.js (hunk 1 above)
Comment (franck-eyraud) [reply src/backgroundScripts/background.js:102]: Done in b1d41be

**Comment by benel** [comment]
> Were you getting it when loading the highlights ?

As I wrote earlier, it is on password protected pages in Cassandre. The exception occurs because API URI containing the same UUID are also password protected.

If you want to test it yourself (before and after upgrading Hypertopic/lib-node), use the URI mentioned in the thread (with your own Hypertopic LDAP login and password, and just ask me for a reset if you don't remember it).

> However, if one is not identified in the argos server, then any modification is currently silently ignored (except the 401 unauthorized displayed in the web browser console), and there it would be probably nice to display some message. Do you have some idea where ?

There is a difference indeed between:
- exceptions on `getView` (which could be highly distributed similarly to a P2P network), that should remain silent,
- exceptions on `put` and `delete` (often preceded by `get`) that means that what the user wanted to do coud not be achieved.

For the latter, there should be indeed a feedback to the user (with of course a single notification by unachieved feature and a message as understandable as possible).

**Comment by franck-eyraud** [comment]
Thanks for the merge!

> For the latter, there should be indeed a feedback to the user (with of course a single notification by unachieved feature and a message as understandable as possible).

I tried to do that for now on the two write requests of this merge (create and remove highlight) in this pull request : https://github.com/Hypertopic/LaSuli/pull/63

**Pull request merged**